server/genquery2/src/genquery2_sql.cpp

```diff
@@ -0,0 +1,1172 @@
+
+                if (const auto iter = state.table_aliases.find(std::string{_t}); iter != std::end(state.table_aliases))
+                {
+                    alias = iter->second;
+                }
+
+                log_gq::debug("TABLE => {} [alias={}]", _t, alias);
+            });
+
+            log_gq::debug("Requires metadata table joins for R_DATA_MAIN? {}", state.add_joins_for_meta_data);
+            log_gq::debug("Requires metadata table joins for R_COLL_MAIN? {}", state.add_joins_for_meta_coll);
+            log_gq::debug("Requires metadata table joins for R_RESC_MAIN? {}", state.add_joins_for_meta_resc);
+            log_gq::debug("Requires metadata table joins for R_USER_MAIN? {}", state.add_joins_for_meta_user);
+            log_gq::debug("Requires table joins for DATA_RESC_HIER? {}", state.add_sql_for_data_resc_hier);
+
+            //
+            // Generate SQL
+            //
+
+            log_gq::trace("### PHASE 2: SQL Generation");
+
+            auto graph = init_graph();
+
+            // Generate the SELECT clause.
+            //
+            // This step does not concern itself with special columns (e.g. META_DATA_ATTR_NAME). Those
+            // will be handled in a later step.
+            //
+            // The tables stored in sql_tables must be directly joinable to at least one other table in
+            // the sql_tables list. This step is NOT allowed to introduce intermediate tables.
+            auto select_clause =
+                fmt::format("{with_clause}select {distinct}{columns} from {table} {alias}",
+                            fmt::arg("with_clause", generate_with_clause_for_data_resc_hier(state, _opts.database)),
+                            fmt::arg("distinct", _select.distinct ? "distinct " : ""),
+                            fmt::arg("columns", cols),
+                            fmt::arg("table", state.sql_tables.front()),
+                            fmt::arg("alias", state.table_aliases.at(std::string{state.sql_tables.front()})));
+
+            log_gq::debug("SELECT CLAUSE => {}", select_clause);
+
+            const auto inner_joins = generate_inner_joins(graph, state.sql_tables, state.table_aliases);
+
+            std::for_each(std::begin(inner_joins), std::end(inner_joins), [](auto&& _j) {
+                log_gq::debug("INNER JOIN => {}", _j);
+            });
+
+            if (inner_joins.size() != state.sql_tables.size() - 1) {
+                throw std::invalid_argument{"invalid general query"};
+            }
+
+            auto sql = select_clause;
+
+            if (!inner_joins.empty()) {
+                sql += fmt::format(" {}", fmt::join(inner_joins, " "));
+            }
+
+            // Q. Should tickets be scoped to data objects and collections separately?
+            // Q. What happens if a user attempts to query data objects, collections, and tickets in the same query?
+            // Q. Should these questions be handled by specific queries instead?
```

Comment:
Got issue(s) to link here?

Reply:
Those will appear after this PR is merged (so I can link to the lines directly).